File: scala/src/main/scala/euler/problem/Problem4.scala
package euler.problem

import euler.util.palindrome.palindromeInt

object Problem4 extends App {

  lazy val answer: Int =
    (1 to 999)
      .map(BigInt(_))
      .combinations(2)
      .map(x => x(0) * x(1))
      .filter(palindromeInt(_))
      .max
      .intValue

}

File: haskell/src/Euler/Problem/Problem4.hs
module Euler.Problem.Problem4 (answer) where

import Data.Sequence (replicateM)
import Data.Foldable (toList)

answer :: Integer
answer = maximum $ filter isPalindrome $ map product $ pairsOf [1..999]
  where
    isPalindrome i = s == reverse s where s = show i
    pairsOf xs = map toList $ replicateM 2 xs
